### PR TITLE
Add to calendar

### LINKED ISSUE #8: Build outgoing cookie logic

### PR DESCRIPTION
**Pull request by hn275**
# Cookie reader/writer

I vaguely remember us discussing that the cookie should store the state in the form of base64-encoded json, this PR implements all the necessary traits for `UserState` to serialize/deserialize from/into `axum_extra::extract::cookie::Cookie` struct:
- `TryFrom`
- `TryInto`
- `Default` <- default the selection `Vec<ThinCourse>` to be empty.

The middleware `parse_cookie` is fixed to use these trait impl's and to properly read the current user state.

## Example: to set a cookie

```rs
let cookie: Cookie = user_state.try_into()?;
let header = AppendHeaders([(SET_COOKIE, cookie.to_string())]);

Ok((header, html! {"valid html string here"}))
```

see full example at [`add_to_calendar`](https://github.com/brennanmcmicking/scheduler/blob/f0a754ce53cb915ee2faa3bc477715320322e6da/src/routes/calendar.rs#L31).

# Add To Calendar API

The endpoint will query the database for the requested `crn` and `term`:

```sh
curl \
-X PUT "http://localhost:8080/calendar" \
-H "Content-Type: application/x-www-form-urlencoded" \
-d "crn=23962&term=202501"
```

If the queried course is already in the set cookie, it does nothing to the cookie's value.
otherwise it pushes the `subject_code` and `course_code` into the current user's selection before setting a new cookie value.

# Bonus

If you have [cargo watch](https://github.com/watchexec/cargo-watch) installed, there's now an alias for hot reloading

```sh
cargo w
```

close #8

### CODE REVIEW

File: src/middlewares.rs
Comment (brennanmcmicking) [inline src/middlewares.rs:35]: Can you please create a ticket for this TODO and add the ticket number here for back-reference?

File: src/routes/calendar.rs
Comment (malcolmseyd) [inline src/routes/calendar.rs]: Please use tracing events intead, like `debug!(?err)`. It adds additional context to the console like the current request and function parameters.

File: src/routes/calendar.rs
Comment (brennanmcmicking) [inline src/routes/calendar.rs]: I know that this is an acceptable method of handling cookies but I would prefer you use the [CookieJar](https://docs.rs/axum-extra/latest/axum_extra/extract/cookie/struct.CookieJar.html) abstraction as it allows our implementation to be a bit more blind to the inner workings of HTTP.

File: src/routes/calendar.rs
Comment (brennanmcmicking) [inline src/routes/calendar.rs:68]: I didn't know that Rust allows you to redefine a variable as mutable within a specific scope.

File: src/middlewares.rs
Comment (malcolmseyd) [inline src/middlewares.rs]: Just curious, when can this come up? I'm not sure that it should be possible to create a `UserState` that can't be serialized, as that would constitute a bug in our program, so I don't think returning a 400 status code is appropriate here.

The two options I'd recommend are either set `type Error = anyhow::Error` and consider this a fatal unhandleable error, or just use `.expect()` to unwrap the error and panic otherwise.

It seems like this would essentially be a bug in our program if we can't serialize the user state, so maybe we should use `.expect()` here and panic instead.

The second options is personally appealing because it would be incredibly convenient to use with CookieJar, since [`CookieJar::add`](https://docs.rs/axum-extra/latest/axum_extra/extract/cookie/struct.CookieJar.html#method.add) takes an `Into<Cookie>`, but idk, it might be good to catch the error instead, your call. I'm still personally trying to decide when it's appropriate to panic or not.

File: src/routes/calendar.rs
Comment (brennanmcmicking) [inline src/routes/calendar.rs]: Please leverage the [CookieJar](https://docs.rs/axum-extra/latest/axum_extra/extract/cookie/struct.CookieJar.html) mechanism rather than setting a header explicitly. I would rather use one greater layer of abstraction.

File: .cargo/config.toml
Comment (malcolmseyd) [inline .cargo/config.toml]: Please put this in `~/.cargo/config.toml` instead, I don't see the reason to share this alias with everyone else.

**Review by malcolmseyd — COMMENTED** [review]
Looks pretty good, just some nitpicks here and there.

File: src/middlewares.rs
Comment (malcolmseyd) [inline src/middlewares.rs]:
```suggestion
            debug!(?err);
```

File: src/routes/calendar.rs
Comment (malcolmseyd) [inline src/routes/calendar.rs]:
```suggestion
        debug!(?err);
```

File: src/middlewares.rs
Comment (hn275) [reply src/middlewares.rs]: I ended up switching the trait to from `TryInto` -> `Into`

**Review by malcolmseyd — COMMENTED** [review]
Looks pretty good! Please fix the last comment about `dbg!` and then I'll be good to approve!

File: src/routes/calendar.rs
Comment (brennanmcmicking) [inline src/routes/calendar.rs]: There is no need to set the user state cookie again if nothing has changed.

File: src/middlewares.rs
Comment (brennanmcmicking) [inline src/middlewares.rs:63]: This should not return a status code because it does not live inside of a request handler. What if I want to use this code somewhere else? What if there is a way to recover gracefully from a cookie parse failure?

File: src/middlewares.rs
Comment (brennanmcmicking) [inline src/middlewares.rs]: For the same reason stated below, this shouldn't be using status code at all IMO.

**Review by brennanmcmicking — CHANGES_REQUESTED** [review]

File: src/middlewares.rs
Comment (brennanmcmicking) [reply src/middlewares.rs]: I suspect you should be using  `Anyhow` instead?

File: src/routes/calendar.rs
Comment (brennanmcmicking) [reply src/routes/calendar.rs:68]: I suppose you're cloning it to owned so nothing of concern I just expected to see a different variable name. Not expecting you to change anything before I approve.

File: src/middlewares.rs
Comment (hn275) [reply src/middlewares.rs:63]: this is a middleware and shouldn't be called outside the context of it being a middleware.

I'm returning a status code since (imo) any ops after that failure shouldn't be allowed, since it seems like a lot of our actions rely on the fact that the user state is valid

File: src/middlewares.rs
Comment (hn275) [reply src/middlewares.rs:35]: https://github.com/brennanmcmicking/scheduler/issues/23 :+1:

File: src/middlewares.rs
Comment (brennanmcmicking) [reply src/middlewares.rs:63]: Does axum expect the middleware to return a status code and will it short-circuit in that case? If so, please link the relevant documentation showing that dealing with status codes in middleware is a blessed path.

Either way, this function is supposed to parse the cookie or fail graciously. If the error can be caught higher up in the call stack and turned into a status code then that should be done instead.

Comment (hn275) [reply src/middlewares.rs:63]: sure, this is the example I based my code off.
https://docs.rs/axum/latest/axum/middleware/index.html#passing-state-from-middleware-to-handlers

Comment (brennanmcmicking) [reply src/middlewares.rs:63]: Ok, I think I understand. If the extension returns an `Err` then axum will never call the request handler.

**Review by brennanmcmicking — APPROVED** [review]

**Comment by malcolmseyd** [comment]
Uh, I don't think this code works. I get a 422 whenever I try clicking a button in the web UI
```
2024-08-07T19:24:49.514612Z DEBUG request{method=PUT uri=/calendar}: tower_http::trace::on_request: started processing request
2024-08-07T19:24:49.514989Z TRACE request{method=PUT uri=/calendar}: axum::rejection: rejecting request status=422 body="Failed to deserialize form body: missing field `term`" rejection_type="axum::extract::rejection::FailedToDeserializeFormBody"
2024-08-07T19:24:49.515159Z DEBUG request{method=PUT uri=/calendar}: tower_http::trace::on_response: finished processing request latency=654 μs status=422
```
My bad for approving without manually testing the code

**Pull request merged**
